third_party/nvfuser/csrc/scheduler/registry.cpp

```diff
@@ -2038,10 +2044,272 @@ class PersistentKernelScheduler : public SchedulerEntry {
   }
 };
 
+class MatmulScheduler : public SchedulerEntry {
+ public:
+  explicit MatmulScheduler(
+      Fusion* fusion,
+      SchedulerRuntimeInfo& runtime_info,
+      HeuristicSummary* data_cache = nullptr)
+      : SchedulerEntry(ScheduleHeuristic::Matmul) {
+    computeHeuristics(fusion, runtime_info);
+  }
+
+  void schedule(Fusion* fusion) override {
+    FUSER_PERF_SCOPE("Schedule Matmul Fusion");
+    scheduleMatmul(fusion, matmulParams());
+  }
+
+  static bool canScheduleCompileTime(Fusion* fusion) {
+    // Check that inputs of all select/gather-like ops are fusion inputs
+    if (rejectScheduleForSelectLikeOps(fusion, ScheduleHeuristic::Matmul)) {
+      return false;
+    }
+
+    // The plan:
+    // 1. check if MmaOp is defined
+    // 2. check if TV bound to MmaOp is in outputs
+    // 3. check if inputs for output are in local memory
+    // 4. check if local_mem of MmaOp have supported layout
+    // 5. [TODO] check if local_mem of MmaOp are a result of broadcast
+    //    on fusion inputs
+
+    // #1
+    std::vector<Expr*> mmaExprs;
+    {
+      // TODO: outline to ir_utils namespace
+      for (const auto expr : fusion->exprs()) {
+        if (expr->isA<MmaOp>()) {
+          mmaExprs.push_back(expr);
+        }
+      }
+    }
+
+    if (mmaExprs.empty()) {
+      scheduler_debug_utils::canScheduleRejectReason(
+          ScheduleHeuristic::Matmul, "No mma op to schedule");
+      return false;
+    }
+
+    // #2
+    // NOTE: this may change in future, with support for C matrix and beta
+    {
+      // TODO: outline to ir_utils namespace
+      const auto& outputs = fusion->outputs();
+      std::vector<Expr*> mmaOutputs;
+
+      for (auto expr : mmaExprs) {
+        const auto& exprOutputs = expr->outputs();
+        if (exprOutputs.size() != 1) {
+          // we are interested only in mma ops with single output val
+          continue;
+        }
+        if (std::find(outputs.cbegin(), outputs.cend(), exprOutputs.front()) !=
+            outputs.cend()) {
+          mmaOutputs.push_back(expr);
+        }
+      }
+
+      mmaExprs = mmaOutputs;
+    }
+
+    if (mmaExprs.empty()) {
+      scheduler_debug_utils::canScheduleRejectReason(
+          ScheduleHeuristic::Matmul, "No output mma op to schedule");
+      return false;
+    }
+
+    // #3
+    {
+      // TODO: outline to ir_utils namespace
+      for (const auto* mmaExpr : mmaExprs) {
+        const auto& mmaExprInputs = mmaExpr->inputs();
+        if (mmaExprInputs.size() != 2) {
+          scheduler_debug_utils::canScheduleRejectReason(
+              ScheduleHeuristic::Matmul,
+              "One of mma op doesn't have 2 inputs, got: " +
```

Comment:
Can the mma op be defined without two input tensor views?

Reply:
That's a good question. I can think of situation where the same TV is used for both inputs but currently `scheduleMartmul` does not support this configuration. My idea for now is to have a skeleton that matches what is expected as of today in matmul tools and extend it when more details are available.

Reply:
I don't think these checks are needed since mma op requires explicitly two inputs one output all TensorViews: https://github.com/csarofeen/pytorch/blob/devel/third_party/nvfuser/csrc/ir_nodes.cpp#L1324-L1336

Reply:
Thanks for pointer to the definition of mma ir node, I will remove this check from here as it is redundant at this point.